MIPs/mip-3.md

```diff
@@ -0,0 +1,184 @@
+
+  ```markdown
+  if (from.length > 1) {
+  throw ethErrors.rpc.invalidParams(
+  'Currently we de not support multiple tokens swap',
+  );
+  }
+  ```
+
+- **Inactive or Unsupported Swaps**: If the swap is inactive or not possible on the current chain, an alert with the message 'Swap is not active or not possible on this chain' will be triggered.
+
+  ```markdown
+  const isSwappable = isSwapsAllowed(chainId) && swapsIsLive;
+  if (!isSwappable) {
+  Alert.alert('Swap is not active or not possible on this chain');
+  return;
+  }
+  ```
+
+5. Security: Dapp Developers should ensure that the token addresses and chain IDs provided in the method call are valid and secure. They should also inform users about the potential risks and considerations when performing token swaps.
+   On MetaMask platfors we have warnings in place that say to the user to be carefull when a token is not that trustworthy. We check if the number of occurrences is more than one, If not we show an alert the we obligate the user to ready and press a button to be able to swap.
+
+6. Testing: Before deploying the `wallet_swapAsset` method in a live environment, MetaMask engineers should thoroughly test it to ensure it works correctly and handles errors appropriately. This includes testing with different token types, amounts, and network conditions.
+
+7. Interoperability: The `wallet_swapAsset` method provides a standardized way for dApps to request token swaps from wallets. This can improve interoperability between different dApps and wallets, as they can all use the same method for token swaps.
+
+## User Experience Considerations
+
+Users will now be able to swap tokens without leaving dapps.
+The `wallet_swapAsset` method might be new to some users, providing educational resources can help them understand how it works and how to use it.
+If the `wallet_swapAsset` method encounters an error, the wallet/dapp should provide a clear and understandable error message to the user. This helps users understand what went wrong and how to fix it.
+
+On MetaMask Mobile, if the user have on the wallet an selected account but on the dapp in app browser another, when using swaps feature, MetaMask Mobile will check if the accounts are different and switch to the account selected on the dapp if they are.
+
+```markdown
+if (
+safeToChecksumAddress(selectedAddress) !==
+checksummedDappConnectedAccount
+) {
+Engine.context.PreferencesController.setSelectedAddress(
+checksummedDappConnectedAccount,
+);
+}
+```
+
+## Security Considerations
+
+While the proposed implementation of `wallet_swapAsset` do not introduce new security risks in terms of contract interactions, they do present a potential risk related to phishing, althought it is good to mention that this risk already exist on our swaps flow.
```

Comment:
```suggestion
While the proposed implementation of `wallet_swapAsset` does not introduce new security risks in terms of contract interactions, they do present a potential risk related to phishing.
```

Not sure if we should mention that last part `althought it is good to mention that this risk already exist on our swaps flow.`

Reply:
yeah will remove it!! I don't think it should me mentioned as well!